url969d1ddca4_compmstr_java_scheme_tgz-pJ8-com_undi_javascheme_JavaSchemeJ8/src/com/undi/util/Reflector.java

```diff
@@ -0,0 +1,371 @@
+    Method m = null;
+    Object[] boxedArgs = null;
+    if(methods.isEmpty()){
+      throw new IllegalArgumentException("Method: '" + methodName + "' does not exist in: " + target);
+    }else if(methods.size() == 1){
+      m = (Method) methods.get(0);
+      boxedArgs = boxArgs(m.getParameterTypes(), args);
+    }else{ //overloaded with same arity
+      Method foundm = null;
+      for(Method method : methods){
+        Class<?>[] params = method.getParameterTypes();
+        if(isCongruent(params, args)){
+          if(foundm == null){
+            foundm = method;
+            boxedArgs = boxArgs(params, args);
+          }
+        }
+      }
+      m = foundm;
+    }
+    if(m == null){
+      throw new IllegalArgumentException("No matching method found");
+    }
+    if(!Modifier.isPublic(m.getDeclaringClass().getModifiers())){
+      //public method of non-public class...
+      //Method oldm = m;
+      m = getAsMethodOfPublicBase(m.getDeclaringClass(), m);
+      if(m == null){
+        throw new IllegalArgumentException("Can't call public method of non-public class");
+      }
+    }
+    try{
+      return prepRet(m.getReturnType(), m.invoke(target, boxedArgs));
+    }catch(Exception e){
+      throw SneakyThrow.sneakyThrow(e);
+    }
+  }
+  
+  public static Object prepRet(Class<?> c, Object obj){
+    if(!(c.isPrimitive() || c == Boolean.class)){
+      return obj;
+    }
+    if(obj instanceof Boolean){
+      return ((Boolean) obj) ? Boolean.TRUE : Boolean.FALSE;
+    }
+    return obj;
+  }
+  
+  static Object boxArg(Class<?> paramType, Object arg){
+    if(!paramType.isPrimitive()){
+      return paramType.cast(arg);
+    }else if(paramType == boolean.class){
+      return Boolean.class.cast(arg);
+    }else if(paramType == char.class){
+      return Character.class.cast(arg);
+    }else if(arg instanceof Number){
+      Number n = (Number) arg;
+      if(paramType == int.class)
+        return n.intValue();
+      else if(paramType == float.class)
+        return n.floatValue();
+      else if(paramType == double.class)
+        return n.doubleValue();
+      else if(paramType == long.class)
+        return n.longValue();
+      else if(paramType == short.class)
+        return n.shortValue();
+      else if(paramType == byte.class)
+        return n.byteValue();
+    }
+    throw new IllegalArgumentException("Unexpected param type, expected: " + paramType +
+        ", given: " + arg.getClass().getName());
+  }
+  
+  static Object[] boxArgs(Class<?>[] params, Object[] args){
+    if(params.length == 0)
+      return null;
+    Object[] ret = new Object[params.length];
+    for(int i = 0; i < params.length; i++){
+      Object arg = args[i];
+      Class<?> paramType = params[i];
+      ret[i] = boxArg(paramType, arg);
+    }
+    return ret;
+  }
+  
+  public static List<Method> getMethods(Class<?> c, int arity, String name, boolean getStatics){
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 15. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test.

We recommend that you simplify this method or break it into multiple methods. For example, consider extracting the code blocks on lines 298-317, lines 318-327 into separate methods.